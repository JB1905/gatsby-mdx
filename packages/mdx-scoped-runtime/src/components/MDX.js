// Direct copy-paste from https://github.com/mdx-js/mdx/releases/tag/v0.20.4
// + fix for object spreading
import React from 'react';
import { transform } from 'buble';
import mdx from '@mdx-js/mdx';
import { MDXTag } from '@mdx-js/tag';

export default ({
  scope = {},
  components = {},
  mdPlugins = [],
  hastPlugins = [],
  children,
  ...props
}) => {
  const fullScope = {
    MDXTag,
    components,
    props,
    ...scope,
  };

  const jsx = mdx
    .sync(children, {
      mdPlugins,
      hastPlugins,
      skipExport: true,
    })
    .trim();

  const { code } = transform(jsx, {
    objectAssign: 'Object.assign',
  });

  const keys = Object.keys(fullScope);
  const values = Object.values(fullScope);
  // eslint-disable-next-line no-new-func
  const fn = new Function(
    '_fn',
    'React',
    ...keys,
    `${code}

  return React.createElement(MDXContent, { components, ...props });`
  );

  return fn({}, React, ...values);
};
