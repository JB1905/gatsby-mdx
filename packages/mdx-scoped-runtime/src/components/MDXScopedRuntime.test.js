import React from 'react';
import { renderToString } from 'react-dom/server';
import prettier from 'prettier';
import MDX from './MDXScopedRuntime';

it('should render the example', () => {
  expect(
    prettier.format(
      renderToString(
        <MDX
          scope={{
            Demo: () => <code>dem0</code>,
            // eslint-disable-next-line react/prop-types
            Layout: ({ children }) => <div id="layout">{children}</div>,
          }}
        >
          {`
import Demo from '../ui/Demo';
import Layout from '../ui/Layout';

export default Layout

# Hello, world!

<Demo />
`}
        </MDX>
      ),
      { parser: 'html' }
    )
  ).toMatchSnapshot();
});

it('should render simple mdx with layout', () => {
  expect(
    prettier.format(
      renderToString(
        <MDX>
          {`
export default ({children}) => <div id="layout">{children}</div>

# Hi
`}
        </MDX>
      ),
      { parser: 'html' }
    )
  ).toMatchSnapshot();
});
