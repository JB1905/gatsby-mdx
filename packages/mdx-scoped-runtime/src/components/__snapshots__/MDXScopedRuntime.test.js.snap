// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render simple mdx with layout 1`] = `
"<div id=\\"layout\\">
  <div><h1>Hi</h1></div>
</div>
"
`;

exports[`should render the example 1`] = `
"<div id=\\"layout\\">
  <div>
    <h1>Hello, world!</h1>
    <code>dem0</code>
  </div>
</div>
"
`;
